### PR TITLE
feat: Add comprehensive test suite for DestChain Contracts, SourceChain Contracts and Token Contract

### LINKED ISSUE #24: Adding test case or scripts for smart contracs
Need to write more test cases, scripts, or tasks to test functions in onramp contracts. 
For example:
- Onramp.sol
  - offerData()
  - commitAggregate()
  - verifyDataStored()
  - more ...
- Prover-Axelar.sol
  -  call_axelar()
  -  more ...

### PR DESCRIPTION
**Pull request by paschal533**
## Description
This PR adds a comprehensive test suite for:
- DestChain Contracts.
- SourceChain Contracts.
- Token Contract.

## Changes
- Added TypeScript tests using Hardhat and Ethers v6.
- Covered all core functionality, edge cases, and error handling.
- Ensured compatibility with Filecoin Calibration testnet and Axelar bridge.

## Testing
Run tests with:
```bash
npx hardhat test --network hardhat
```

Closes #24

### CODE REVIEW

**Comment by longfeiWan9** [comment]
@paschal533 can you please resolve the conflicts?

**Comment by paschal533** [comment]
Hi @longfeiWan9 , I have resolved the conflicts

File: test/destChain_test/Prover_Axelar_test.ts [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+    it("Should set destination chains correctly", async function () {
+      const chainIds = [1, 2, 3];
+      const chainNames = ["ethereum", "polygon", "avalanche"];
+      const addresses = [addr1.address, addr2.address, owner.address];
+
+      await dealClient.setDestinationChains(chainIds, chainNames, addresses);
+
+      const chain1 = await dealClient.chainIdToDestinationChain(1);
+      expect(chain1.chainName).to.equal("ethereum");
+      expect(chain1.destinationAddress).to.equal(addr1.address);
```
Comment (longfeiWan9) [inline test/destChain_test/Prover_Axelar_test.ts:103]: can you also verify if there are 3 chains to be added?

File: test/destChain_test/Prover_Axelar_test.ts [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+
+      await expect(
+        dealClient.handle_filecoin_method(unhandledMethod, 0n, emptyParams)
+      ).to.be.revertedWith(
+        "the filecoin method that was called is not handled"
+      );
+    });
+  });
+
+  describe("AxelarBridge", function () {
```
Comment (longfeiWan9) [inline test/destChain_test/Prover_Axelar_test.ts]: Is this for AxelarBridge contract from [Oracle.sol](https://github.com/FIL-Builders/onramp-contracts/blob/main/contracts/sourceChain/Oracles.sol)? Any specific reason you add the test cases here?

File: test/destChain_test/Prover_Axelar_test.ts [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+      const addresses = [addr1.address, addr2.address, owner.address];
+
+      await dealClient.setDestinationChains(chainIds, chainNames, addresses);
+
+      const chain1 = await dealClient.chainIdToDestinationChain(1);
+      expect(chain1.chainName).to.equal("ethereum");
+      expect(chain1.destinationAddress).to.equal(addr1.address);
+    });
+
+    it("Should handle filecoin methods correctly", async function () {
```
Comment (longfeiWan9) [inline test/destChain_test/Prover_Axelar_test.ts:114]: can you also test the `handle_filecoin_method` for `MARKET_NOTIFY_DEAL_METHOD_NUM` ? This is a very important operation for storage deal flow. It should invoke `dealNotify(params)`.

File: test/destChain_test/Prover_Axelar_test.ts [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+      await axelarBridge.setSenderReceiver(
+        addr1.address,
+        await debugReceiver.getAddress()
+      );
+
+      // Convert address to string format expected by Axelar
+      const sourceAddressString = addr1.address.toLowerCase();
+
+      // Call the contract using the execute function from base contract
+      await expect(
```
Comment (longfeiWan9) [inline test/destChain_test/Prover_Axelar_test.ts:162]: `AxelarBridge._execute()` is implemented to overide `_execute()` in `AxelarExecutable` so that Axelar Bridge can call this function when there is a cross-chain invocation coming from prover-axelar.sol. Do we need to test it with the Axelar bridge?

File: test/destChain_test/Prover_Axelar_test.ts [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+          dealId: 1n,
+        };
+
+        const params = Buffer.from(cbor.encode(marketDealNotifyParams));
+
+        // Impersonate market actor and notify deal
+        const impersonatedSigner = await ethers.getImpersonatedSigner(
+          MARKET_ACTOR_ADDRESS
+        );
+        await owner.sendTransaction({
```
Comment (longfeiWan9) [inline test/destChain_test/Prover_Axelar_test.ts:311]: What does this function do? I do not see how is it calling dealNotify()

**Comment by snissn** [comment]
This looks generally good to me! I haven't dived too deep into each test yet, but it's a huge improvement and step forward from not having any tests! 

Also added a github workflow to run the hardhat tests on push and PR - @paschal533 please merge my PR into your branch before merging into this repo -- https://github.com/paschal533/onramp-contracts/pull/1

**Comment by snissn** [comment]
@paschal533 are you able to address @longfeiWan9's questions? Thanks in advanced!

**Comment by paschal533** [comment]
> [test/destChain_test/Prover_Axelar_test.ts](/FIL-Builders/onramp-contracts/pull/26/files/47cd9aa4dbaf6bb4417080b0ac7c96ea9c4dd0df#diff-af6b7c35bcdf95768435fefaa6afec1eaa472f9c5c5e61ce5af519ca17fb12f5)
>         const impersonatedSigner = await ethers.getImpersonatedSigner(



> @paschal533 are you able to address @longfeiWan9's questions? Thanks in advanced!

Hi @snissn, I will address and resolve them by tomorrow. So sorry for the delay. I was a bit busy.

**Comment by paschal533** [comment]
> This looks generally good to me! I haven't dived too deep into each test yet, but it's a huge improvement and step forward from not having any tests!
> 
> Also added a github workflow to run the hardhat tests on push and PR - @paschal533 please merge my PR into your branch before merging into this repo -- [paschal533#1](https://github.com/paschal533/onramp-contracts/pull/1)

Thank you for the kind words. I really appreciate it.

**Comment by snissn** [comment]
> > [test/destChain_test/Prover_Axelar_test.ts](/FIL-Builders/onramp-contracts/pull/26/files/47cd9aa4dbaf6bb4417080b0ac7c96ea9c4dd0df#diff-af6b7c35bcdf95768435fefaa6afec1eaa472f9c5c5e61ce5af519ca17fb12f5)
> > const impersonatedSigner = await ethers.getImpersonatedSigner(
> 
> > @paschal533 are you able to address @longfeiWan9's questions? Thanks in advanced!
> 
> Hi @snissn, I will address and resolve them by tomorrow. So sorry for the delay. I was a bit busy.

Amazing! Thanks for all your hard work!

File: test/destChain_test/Prover_Axelar_test.ts (hunk 1 above)
Comment (paschal533) [reply test/destChain_test/Prover_Axelar_test.ts:103]: Sure, let me do that right away. Thanks for the feedback

File: test/destChain_test/Prover_Axelar_test.ts (hunk 2 above)
Comment (paschal533) [reply test/destChain_test/Prover_Axelar_test.ts]: Ah, good catch! I initially wrote all tests in a single file but later moved them to separate files for better organization. This AxelarBridge test was meant to be in the Oracle test file but I must have missed moving it during the refactoring. I'll move it to the Oracle test file right away. Thanks for catching this

File: test/destChain_test/Prover_Axelar_test.ts (hunk 3 above)
Comment (paschal533) [reply test/destChain_test/Prover_Axelar_test.ts:114]: I'm still running into an issue with the MARKET_NOTIFY_DEAL_METHOD_NUM test. Even after several attempts at fixing the CBOR encoding, I'm getting the error "invalid maj (expected MajUnsignedInt)". I've tried different CBOR structures but haven't been able to resolve it. Could you help me understand the correct CBOR structure that the contract expects for the deal parameters? I want to make sure we have proper test coverage for this critical flow.

File: test/destChain_test/Prover_Axelar_test.ts (hunk 5 above)
Comment (paschal533) [reply test/destChain_test/Prover_Axelar_test.ts:311]: The impersonatedSigner part is used to simulate a call from the Market Actor (0xff00000000000000000000000000000000000005) since dealNotify() can only be called via handle_filecoin_method by the Market Actor. The test tries to create a deal proposal and call handle_filecoin_method with MARKET_NOTIFY_DEAL_METHOD_NUM (4186741094), which internally calls dealNotify().

However, I notice two issues in my test:
1. The crucial part that actually calls dealNotify() is commented out (the handle_filecoin_method call)
2. The test is still failing due to CBOR encoding issues that I haven't resolved yet

File: test/destChain_test/Prover_Axelar_test.ts (hunk 3 above)
Comment (longfeiWan9) [reply test/destChain_test/Prover_Axelar_test.ts:114]: How about trying to verify if DealNotify will be called in handle_filecoin_method? I am also not familiar with CBOR as well, maybe we can pass the params part first.

Comment (paschal533) [reply test/destChain_test/Prover_Axelar_test.ts:114]: Alright, I will give that a try and keep you posted. Thank you.

**Comment by paschal533** [comment]
Hi @longfeiWan9, I'm currently making some changes in the testing (kind of rewriting the whole test). I will be done by tomorrow and let you know.

**Comment by longfeiWan9** [comment]
@paschal533 hi, just want to follow up about the progress.

**Comment by paschal533** [comment]
Hi @longfeiWan9, Thanks for following up on the progress. I’ve made some solid headway on this—49 out of 50 tests are now passing, which I’m pretty happy about. You can see the details in the latest test run output, but here’s a quick summary: all the `DealClientAxl`, `OnRampContract`, `Bridge Contracts`, and `ERC20 Token Contracts` tests are passing, including key functionality like `makeDealProposal`, event emissions, and cross-chain flows. the `DealProposalCreate` event test and the deal storage test under `DealClientAxl` are both green now, along with the integration tests.

However, I’m still wrestling with one stubborn test case: `"Should validate market actor address in dealNotify"`. It’s failing with the error `AssertionError: Expected transaction NOT to be reverted, but it reverted with reason 'expected major type tag when parsing cid'`. I’ve tried everything I can think of to fix it—adjusting the CBOR encoding for the `piece_cid` field with and without tag 42, tweaking the CID format, and even manually constructing the `dealProposal` array to match what `filecoin-solidity-api` expects. The latest `Params Hex` looks like this:

```
0x8258888b58220170f470e59cc8f6d99657f6c915716229a03c6edb53b2c0180efe18a5e24a5b1e5d01f45400000000000000000000000000000000000000005820bb3b9641ccaa4395290be800a5fe66183d6040ed006df732df786f6080fb154678420000000000000000000000000000000000000000000000000000000000000001010241014101410101
```

The `piece_cid` is encoded as `0x5822<0170<32-byte hash>>` (untagged), but `readCid` in `CborDecode.sol` expects a tagged CID (`0xd82a5822<34 bytes>`). When I add the tag manually, it still fails, likely due to an extra byte string wrapper I can’t shake off with `cbor.encode`. I’ve commented out this test in my last commit to keep the suite clean while we figure it out, but I’m determined to crack it. I’d really appreciate your input to get this last test passing. Let me know what you think :)
